runtime/compiler/env/DependencyTable.cpp

```diff
@@ -0,0 +1,169 @@
+
+   auto it = _offsetMap.find(classOffset);
+   auto entry = getOffsetEntry(classOffset, false);
+   if (entry)
+      entry->_loadedClasses.erase((J9Class *)clazz);
+   }
+
+// If an entry exists for a class, remove it. Otherwise, if we should
+// revalidate, add an entry if the class has a valid chain.
+void
+TR_AOTDependencyTable::recheckSubclass(J9Class *ramClass, uintptr_t offset, bool shouldRevalidate)
+   {
+   auto it = _offsetMap.find(offset);
+   if (it != _offsetMap.end())
+      {
+      auto c_it = it->second._loadedClasses.find(ramClass);
+      if (c_it != it->second._loadedClasses.end())
+         it->second._loadedClasses.erase(c_it);
+      else if (_sharedCache->classMatchesCachedVersion(ramClass, NULL))
+         it->second._loadedClasses.insert(c_it, ramClass);
+      }
+   else if (shouldRevalidate && _sharedCache->classMatchesCachedVersion(ramClass, NULL))
+      {
+      PersistentUnorderedSet<J9Class *> loadedClasses(PersistentUnorderedSet<J9Class *>::allocator_type(TR::Compiler->persistentAllocator()));
+      loadedClasses.insert(ramClass);
+      _offsetMap.insert(it, {offset, {loadedClasses}});
+      }
+   }
+
+// In a class redefinition event, an old class is replaced by a fresh class. If
+// the ROM class offset changed as a result, it and all its subclasses that
+// formerly had valid chains will now be guaranteed not to match, so the entries
+// for these must be removed. If the new offset is valid, any class that didn't
+// have an entry should be rechecked.
+void
+TR_AOTDependencyTable::invalidateRedefinedClass(TR_PersistentCHTable *table, TR_J9VMBase *fej9, TR_OpaqueClassBlock *oldClass, TR_OpaqueClassBlock *freshClass)
+   {
+   if (!_sharedCache)
+      return;
+
+   uintptr_t freshClassOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   uintptr_t oldClassOffset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+   if (!_sharedCache->isClassInSharedCache(freshClass, &freshClassOffset) && !_sharedCache->isClassInSharedCache(oldClass, &oldClassOffset))
+      return;
+
+   if (oldClassOffset != freshClassOffset)
+      return;
+
+   bool revalidateUntrackedClasses = freshClassOffset != TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+
+   TR_PersistentClassInfo *classInfo = table->findClassInfo(oldClass);
+   TR_PersistentCHTable::ClassList classList(TR::Compiler->persistentAllocator());
+
+   table->collectAllSubClasses(classInfo, classList, fej9);
+   classList.push_front(classInfo);
+
+   OMR::CriticalSection cs(_tableMonitor);
+
+   for (auto iter = classList.begin(); iter != classList.end(); iter++)
+      {
+      auto clazz = (J9Class *)(*iter)->getClassId();
+      uintptr_t offset = TR_SharedCache::INVALID_ROM_CLASS_OFFSET;
+      if (!_sharedCache->isClassInSharedCache(clazz, &offset))
+         continue;
+      recheckSubclass(clazz, offset, revalidateUntrackedClasses);
+      }
+   }
+
+TR_OpaqueClassBlock *
+TR_AOTDependencyTable::findClassCandidate(uintptr_t romClassOffset)
```

Comment:
Is the purpose of this API to materialize a value for say a Profiled Guard test?

Reply:
Yes. I could have put it in a later PR, but I added it as a fallback when a class couldn't be found in a class loader, since that was sometimes happening in my testing. It could be used instead of looking up the class in the loader, I suppose, depending on the efficiency of `jitGetClassInClassloaderFromUTF8`.